donor_assignment/likelihood_per_region.py

```diff
@@ -94,55 +94,60 @@ def main():
     # Generate a DF (umi_probs_position_index) with a probability a umi came
     # from a donor for each barcode-umi combination
     #
-    # umi_probs_position_index ([barcode, umi] x [pos, chr, read, prob_A,
-    # prob_C, prob_G, prob_T]) starts with baseline probability base is an ACTG
-    umi_probs_position_index = single_base_uniq_reads.copy().reset_index('pos')
-    num_donors = len(donors)
-    temp_probs = 0
-    for base in 'ACGT':
-        # Indicator vector of shape [umi_probs_position_index.shape[0] (barcode-umi-pos)] x 1
-        base_is_ref = (genotypes.REF.loc[umi_probs_position_index.pos] == base).values.reshape((-1, 1))
-        base_is_alt = (genotypes.ALT.loc[umi_probs_position_index.pos] == base).values.reshape((-1, 1))
-        base_is_neither = 1 - (base_is_alt + base_is_ref)
-
-        # temp_probs ([pos] x [donor] with positions repeated in umi_probs_position_index ordering)
-        # sum[base] P(donor contributed REF / ALT allele) * Indicator(base is REF / ALT) * P(base | observed bases as SNP)
-        temp_probs += ref_probs.loc[umi_probs_position_index.pos] * base_is_ref * umi_probs_position_index[f'prob_{base}'].values.reshape((-1, 1))
-        temp_probs += alt_probs.loc[umi_probs_position_index.pos] * base_is_alt * umi_probs_position_index[f'prob_{base}'].values.reshape((-1, 1))
-        temp_probs += (1 / num_donors) * base_is_neither * umi_probs_position_index[f'prob_{base}'].values.reshape((-1, 1))
-
-    # umi_probs_position_index ([barcode, umi] x [pos, chr, read, prob_A, prob_C, prob_G, prob_T, donor_1 .... donor_k for k donors])
-    # has the total probability that a barcode/umi came from a donor per donor
-    assert (umi_probs_position_index.pos == temp_probs.index).all()
-    tmp = pd.concat([umi_probs_position_index.reset_index(), temp_probs.reset_index(drop=True)], axis=1)
-    umi_probs_position_index = tmp.set_index("barcode UMI".split())
-    print("UMI per donor probability done")
-
-    #
-    # Regularize the donor probabilities
-    # log transform
-    # Sum log likelihoods per barcode
-    #
-    regularized_log_probs = umi_probs_position_index.copy()
-    regularized_log_probs[donors] *= 0.95
-    regularized_log_probs[donors] += 0.05 / num_donors
-    regularized_log_probs[donors] = np.log(regularized_log_probs[donors])
-
-    barcode_log_likelihood = regularized_log_probs.groupby(['barcode'])[donors].sum()
-    print("Barcode per donor probability done.")
-
-    #
-    # Add in umi counts and snp counts
-    #
-
-    num_snps = umi_probs_position_index.groupby(['barcode']).size()
-    num_umis = umi_probs_position_index.reset_index().groupby('barcode')['UMI'].unique().str.len()
-    barcode_log_likelihood['num_snps'] = num_snps
-    barcode_log_likelihood['num_umis'] = num_umis
-
+    unique_barcodes = np.unique(single_base_uniq_reads.reset_index('barcode').barcode)
```

Comment:
Why np.unique() vs. (...).barcode.unique()

Reply:
It should be the same.  Maybe test to be sure, but if .unique() isn't doing what we expect then there are deeper problems.